### ISSUE TITLE
Feature Request: Allow Users to Set OpenAI and Other Backend APIs/Keys via UI

### ISSUE THREAD

**Issue by ljq29** [hoarder-app/hoarder#492]
#### Description
It would be highly beneficial if the application allowed users to configure the OpenAI API and other backend API keys directly from the user interface (UI) or an admin panel. This would provide more flexibility for users and allow them to use their own API keys.

#### Benefits
- **Customization**: Users can easily use their own API keys, allowing for more tailored use of services like OpenAI and others.
- **Security**: Storing API keys in a secure, user-accessible location would allow users to manage their credentials without modifying the core application.
- **Scalability**: This feature could enable better scalability and flexibility for users who require specific backend services.

#### Proposed Solution
Provide a settings page in the UI where users can input and update their OpenAI API key and other relevant backend service keys. This page should ensure that:
- Keys are stored securely.
- Users can easily update or revoke their API keys if needed.
- There is clear documentation for users on how to input their keys.

#### Additional Context
This feature would streamline setup for users, making it easier to configure backend services without having to modify the source code directly. It would also allow for easy key management in cases where API keys are rotated or need to be updated.

**Comment by kamtschatka** [comment]
Sounds like you mixed together a few things.
* Moving the configurability to the admin page is already planned in https://github.com/hoarder-app/hoarder/issues/280
* Sounds like what you actually want is to have users defined their own API keys (on a user basis instead of global). That would require moving the config to the Settings page, as the admin page is global and can only be accessed by the admin.
* You are saying that you need to update the source code, which is not true, you have to set the environment variables correctly at the moment.

Please update the issue accordingly to what you actually expect.

**Comment by ljq29** [comment]
Description

It would be valuable to allow users to configure their own OpenAI API keys and other backend service keys on a per-user basis directly from the user interface (UI), specifically in the Settings page. This would provide users with the flexibility to input and manage their own API keys without needing to modify global configurations or environment variables.

Benefits

Customization: Users can configure their own API keys, allowing for more personalized access to services like OpenAI and others.
Security: Users can securely store and manage their API keys within the UI, reducing the need to modify environment variables or the core application.
Scalability: Enabling per-user key management allows the application to scale more effectively, supporting users with different needs for backend services.
Proposed Solution

Provide a Settings page where users can input and update their OpenAI API keys and other relevant backend service keys.
Ensure the following:
Keys are stored securely.
Users can update or revoke their API keys with ease.
Provide clear documentation on how to input and manage API keys.
Additional Context

This feature would simplify the user setup process, making it easier to configure backend services without altering environment variables or source code. It would also allow for efficient management in cases where API keys need to be rotated or updated frequently.

Clarification

The goal here is to enable per-user API key management via the Settings page, not to modify the global configuration available in the admin panel. I understand that global settings are already planned for the admin page in [Move the config to the admin page #280](https://github.com/hoarder-app/hoarder/issues/492#280), but this proposal focuses on allowing individual users to manage their own keys.

**Comment by Hyphen-H** [comment]
Yes, it would be really beneficial for areas where ppl can't pay for an official openai apikey due to complicated reasons & scenarios, so adding custom openai-like-api configs in the setting page would be useful.
Here's what I hope that hoarder would support someday❤️
- custom ai-service provider besides openai
- custom api url
- hot reload apikey configs (admin global setting, or per-user setting, of which you see fit, in settings page, without having to modify .env or restarting the container.)
